river/expert/bandit.py

```diff
@@ -0,0 +1,504 @@
+            )
+
+        # Check that the model and the metric are in accordance
+        for model in models:
+            if not metric.works_with(model):
+                raise ValueError(
+                    f"{metric.__class__.__name__} metric can't be used to evaluate a "
+                    + f"{model.__class__.__name__}"
+                )
+        super().__init__(models)
+        self.metric = copy.deepcopy(metric)
+        self._y_scaler = copy.deepcopy(preprocessing.StandardScaler())
+
+        # Initializing bandits internals
+        self._n_arms = len(models)
+        self._n_iter = 0  # number of times learn_one is called
+        self._N = [0] * self._n_arms
+        self.explore_each_arm = explore_each_arm
+        self.average_reward = [0.0] * self._n_arms
+
+        # Warm up
+        self.start_after = start_after
+        self.warm_up = True
+
+        # Randomization
+        self.seed = seed
+        self._rng = random.Random(seed)
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}"
+            + f"\n\t{str(self.metric)}"
+            + f"\n\t{'Best model id: ' + str(self._best_model_idx)}"
+        ).expandtabs(2)
+
+    @abc.abstractmethod
+    def _pull_arm(self):
+        pass
+
+    @abc.abstractmethod
+    def _update_arm(self, arm, reward):
+        pass
+
+    @abc.abstractmethod
+    def _pred_func(self, model):
+        pass
+
+    @property
+    def _best_model_idx(self):
+        # average reward instead of cumulated (otherwise favors arms which are pulled often)
+        return argmax(self.average_reward, self._rng)
+
+    @property
+    def best_model(self):
+        return self[self._best_model_idx]
+
+    @property
+    def percentage_pulled(self):
+        if not self.warm_up:
+            percentages = [n / sum(self._N) for n in self._N]
+        else:
+            percentages = [0] * self._n_arms
+        return percentages
+
+    def predict_one(self, x):
+        best_arm = self._best_model_idx
+        y_pred = self._pred_func(self[best_arm])(x)
+
+        return y_pred
+
+    def learn_one(self, x, y):
+        self._learn_one(x, y)
+        return self
+
+    def add_models(self, new_models: typing.List[base.Estimator]):
+        length_new_models = len(new_models)
+        self.models += new_models
+        self._n_arms += length_new_models
+        self._N += [0] * length_new_models
+        self.average_reward += [0.0] * length_new_models
+
+    def _learn_one(self, x, y):
+        # Explore all arms pulled less than `explore_each_arm` times
+        never_pulled_arm = [i for (i, n) in enumerate(self._N) if n < self.explore_each_arm]
+        if never_pulled_arm:
+            chosen_arm = self._rng.choice(never_pulled_arm)
+        else:
+            chosen_arm = self._pull_arm()
+
+        # Predict and learn with the chosen model
+        chosen_model = self[chosen_arm]
+        y_pred = chosen_model.predict_one(x)
```

Comment:
This might be `predict_proba` for a classifier right?

Reply:
Yes, you're right, didn't anticipate that. Also, for classifier the whole scaling thing is less of a problem (since the target is {0, 1}).